### DISCUSSION TITLE
Running Uvicorn from inside a running loop

### DISCUSSION THREAD

**Discussion by JimiC** [Kludex/uvicorn#2457] [Potential Issue] [answered]
I'm developing an API using FastAPI and Uvicorn as the server runner. Initially my plan is to use a class wrapper around FastAPI and call uvicorn from there. The entire app is designed to be asynchronous.
While on the task I stubbled upon the 'Event loop is already running' issue and when I changed my approach I stumbled upon the 'Event loop is closed' issue.
Looking to find the cause of my issues I looked inside Uvicorn's code and was surprised to see that `asyncio.run()` is called inside `server.run`
 https://github.com/encode/uvicorn/blob/47304d9ae76321f0f5f649ff4f73e09b17085933/uvicorn/server.py#L65
 instead of the entry point 
 https://github.com/encode/uvicorn/blob/47304d9ae76321f0f5f649ff4f73e09b17085933/uvicorn/main.py#L589
 
Is this done by design or for convenience?

**Comment by ForceFledgling** [comment]
The use of asyncio.run() inside the server.run() method in Uvicorn is likely a deliberate decision to simplify cases where the application is entirely asynchronous and doesn't require manual event loop management.

The asyncio.run() function starts a new event loop, runs the coroutine, and closes the loop after the execution is complete. This simplifies the task for standard Uvicorn use cases. However, if your code is already running inside an existing event loop, this can cause the "Event loop is already running" error.

The main goal of this approach is to make Uvicorn usage easier for most users. However, in advanced scenarios, such as when you are already managing the event loop manually or in complex asynchronous systems, you should avoid calling asyncio.run() and work directly with the existing event loops.

**Reply by JimiC** [reply]
@ForceFledgling Thanks for taking the time to respond. 
I would like though an official answer from the development team.

**Comment by Kludex** [comment, marked as answer]
https://www.uvicorn.org/#config-and-server-instances

**Reply by JimiC** [reply]
@Kludex This is already known. The thing is that this way I'm losing all the pre-wiring that `Uvicorn` already does and I will have to duplicate it in my code.

So, the way `Uvicorn` works is indeed by design. The actual entry point then is `server.run`.

As per Python docs,
https://docs.python.org/3/library/asyncio-runner.html#id1
> This function should be used as a main entry point for asyncio programs, and should ideally only be called once.

Thanks again, all.
